biofefi/machine_learning/learner.py

```diff
@@ -37,8 +37,13 @@ def _fit_holdout(self, data: Tuple) -> None:
 
         for i in range(self._opt.n_bootstraps):
             self._logger.info(f"Processing bootstrap sample {i+1}...")
-            X_train, X_test = data.X_train[i], data.X_test[i]
-            y_train, y_test = data.y_train[i], data.y_test[i]
+
+            # Access data using dictionary keys
+            X_train = data.get("X_train", data["X_train"])[i]
+            X_test = data.get("X_test", data["X_test"])[i]
```

Comment:
How can you have `data["X_train"]` as the default for trying to get the key `"X_train"` on `data`? Same for the other keys

Reply:
Yes, Dan, you're absolutely correct! Upon reviewing a test copy of the code, I found that no changes were made to the relevant section. I mistakenly associated the issue with this part of the code due to confusion caused by other errors.

Apologise for the misunderstanding caused.  I've now implemented the changes you suggested. Could you please review the updated code?